### DISCUSSION TITLE
Detect and initiate failover on pgBouncer failure?

### DISCUSSION THREAD

**Discussion by brybicki-cyferd** [patroni/patroni#3210] [Q&A] [answered]
Hi,

Let me start by saying that I absolutely love the Patroni project and appreciate the great work people put into it. Everything works like a charm and makes my life extremely easy. There is a problem in my paradise now though.. since I've been asked to add a connection pooler.

I've been searching the web for two days trying to find a way of maintaining HA while introducing pgBouncer into my setup. Nothing promising tuned up. There are plenty of tutorials that show how to set up Patroni + etcd + HAProxy + pgBouncer but all of them seem to skip over the fact that pgBouncer is a single point of failure.

### Current infra
My setup consists of 3 EC2 instances that all run:
- Patroni
- etcd
- HAProxy
- pgBouncer (new addition)

This is fronted with an **AWS NLB** which:
- Has the 3 aforementioned nodes in its target group.
- Monitors the HAProxy health check ports.
- Can send Postgres traffic to any of the 3 instances as long as HAProxy is healthy.

HAProxies monitor the `8008` ports to discover the Patroni leader. Nodes that don't return a `200` code (all replicas) are flagged as down and all traffic is then forwarded to the pgBouncer port on the leader. pgBouncer forwards the queries to postgres running on the same instance.

So:
- The AWS NLB is highly-available by default.
- Any of the 3 HAProxies is able to handle all requests so 2 of them can die without affecting postgres.
- The etcd cluster is highly-available.
- Patroni will automatically handle **node crashes** and `patroni.service` crashes.

### Problem
pgBouncer **is not** highly-available and if `pgbouncer.service` crashes Patroni won't notice as it's not aware of it. HAProxy will route queries to a dead port.

### Questions
- Is there any recommended way of dealing with this problem?
- Can Patroni be "made aware" that it's running alongside pgBouncer?
- Have I missed something in the pgBouncer setup and it should also be highly-available?

### Potential solutions
1. I'm thinking of enhancing the HAProxy checks to also monitor the pgBouncer port and somehow trigger a failover if it dies on the Patroni leader node. This sounds very hacky and would probably require some scripting and duct tape which I am not a fan of.
2. I stumbled upon some random SO comment suggesting pgBouncer can be skipped if it's down. HAProxy would have to detect this and send the traffic straight to the postgres port. I'm not sure if this is possible to configure.

I wonder if anyone pondered and maybe resolved a similar issue?
All suggestions welcome, and thanks for any help.

**Comment by CyberDem0n** [comment, marked as answer]
I would actually recommend to monitor pgbouncer and restart it if something is wrong.
Doing failover on pgbouncer failure is overkill.

**Reply by brybicki-cyferd** [reply]
Got it. Thank you :)
